rpki-validator/src/main/java/net/ripe/rpki/validator3/domain/validation/CertificateTreeValidationService.java

```diff
@@ -255,170 +255,178 @@ private boolean isValidationRunCompleted(ValidationResult validationResult) {
-                    final ValidationResult vr = tuple.v2();
-                    validateCertificateAuthority(trustAnchor, registeredRepositories, tuple.v1(), vr, accumulator);
-                    return vr;
-                })
-                .forEachOrdered(temporary::addAll);
-        } catch (ManifestEntryException e){
-            temporary.addAll(e.getVr());
-        }
-        catch (Exception e) {
-            validationResult.error(ErrorCodes.UNHANDLED_EXCEPTION, e.toString(), ExceptionUtils.getStackTrace(e));
+
+            Tuple2<ValidationResult, List<CertificateRepositoryObjectValidationContext>> validatedManifestEntries
+                    = validateManifestEntries(trustAnchor, context, manifestUri, manifestObject.get(), manifest, crlUri, x509Crl, accumulator);
+            result.addAll(validatedManifestEntries.v1());
+
+            validatedManifestEntries.v2().stream().map(childContext ->
+                    validateCertificateAuthority(trustAnchor, registeredRepositories, childContext, accumulator)
+            ).forEachOrdered(result::addAll);
+        } catch (Exception e) {
+            result.error(ErrorCodes.UNHANDLED_EXCEPTION, e.toString(), ExceptionUtils.getStackTrace(e));
         }
+        return result;
+    }
 
-        finally {
-            validationResult.addAll(temporary);
+    private Tuple2<ValidationResult, List<CertificateRepositoryObjectValidationContext>> validateManifestEntries(
+            TrustAnchor trustAnchor,
+            CertificateRepositoryObjectValidationContext context,
+            URI manifestUri,
+            RpkiObject manifestObject,
+            ManifestCms manifest,
+            URI crlUri,
+            X509Crl x509Crl,
+            Accumulator accumulator
+    ) {
+        ValidationResult temporary = newValidationResult(context.getLocation());
+        List<ManifestEntryValidationResult> objectValidationResult = manifest.getFiles().entrySet().parallelStream()
+                .map(entry -> validateManifestEntry(trustAnchor, context, manifestUri, crlUri, x509Crl, entry))
+                .collect(toList());
+
+        objectValidationResult.stream().map(ManifestEntryValidationResult::getValidationResult).forEach(temporary::addAll);
+        if (validationConfig.isStrictValidation() && temporary.hasFailures()) {
+            return new Tuple2<>(temporary, Collections.emptyList());
         }
+
+        List<ValidManifestObject> validObjects = objectValidationResult.stream().flatMap(ValidManifestObject::of).collect(toList());
+
+        accumulator.add(manifestObject.key(), manifest, ImmutableSortedSet.of(manifestUri.toASCIIString()));
+        validObjects.forEach(object -> accumulator.add(
+                object.key(),
+                object.getCertificateRepositoryObject(),
+                ImmutableSortedSet.of(object.getLocation().toASCIIString()))
+        );
+
+        List<CertificateRepositoryObjectValidationContext> children = validObjects.stream().flatMap(object -> {
```

Comment:
This one can potentially get pretty big (we have a big flat manifest of the RIPE NCC's production CA). Maybe not a big deal, but if we can avoid instantiating the whole list in memory, would be nice.